Sources/ApodiniDelta/WebServiceStructure/Service.swift

```diff
@@ -0,0 +1,92 @@
+//
+//  File.swift
+//  
+//
+//  Created by Eldi Cano on 22.03.21.
+//
+
+import Foundation
+
+extension Apodini.Operation: ComparableProperty {}
+class ServicePath: PrimitiveValueWrapper<String> {}
+class HandlerName: PrimitiveValueWrapper<String> {}
+
+/// Represents an endpoint
+struct Service: Codable {
```

Comment:
As far as I can see a `Service` is a more compact representation of an `Endpoint` that is used to compare different `Endpoints`. I find the name `Service` not really suiting for this abstraction but most be honest that I don't have a suggestion for a better name at the moment. While I see the appeal for creating a new data structure to compare two different endpoints and their parameters I am a bit concerned that this is going to lead to some code duplication and difficulties when we start to evolve the semantic model including simplifying `Endpoints` in the future.

I would like to get your honest opinion on this: Wouldn't it also be possible to make endpoint conform to `Codable` and `ComparableObject` in an extension as part of this target without creating a new type? I think that cleaning up an `Endpoint` and reducing it to the necessary properties and lazily computing a lot of different properties (e.g. out of the context) might most possible benefit the system as a whole a lot more. Maybe also @lukaskollmer @theMomax and @Supereg 's option on this would be very interesting! I could see that a way to make the existing `WebServiceModel` and `Endpoints` `Codable` would also reduce the need for a dedicated exporter. Not only for your work but also for the work that @lukaskollmer did with this exporter.

Reply:
The same also applies to `ServiceParameter` and `WebServiceStructure`. I think that the current semantic model grew very organically and that focusing on this would be a good think that @theMomax is also addressing. I already noted this in a meeting with him: I would be fine to reduce the number of currently supported exporters to REST/JSON API and an OpenAPI exporter for now if this makes it easier to better structure the WebServiceModel and the complete semantic model and try to integrate gRPC, WebSockets, and GraphQL in a next step but keep them in the back of our heads when improving the inner workings of Apodini.

Reply:
(disclaimer: i only skimmed the code of this PR, so I could be missing something here)

---

@PSchmiedmayer 

I did look into making Apodini's "core" data structures `Codable` as part of my code, it's not really possible.
Some thoughts on why (this is essentially a summary of my reasoning as to why I went with the implementation approach I ended up using)

- Making `Endpoint` conform to `Codable`: this is effectively impossible, because a) it is a generic type, and b) most of its properties cannot be encoded.
- Making the `WebServiceModel` conform to `Codable`:
  - The type itself is not generic, but it is a reference type (which in itself does make not conforming to `Codable` impossible, but greatly complicates things, especially if there's recursion involved, which I don't think is the case here).
  - But: the individual `EndpointTreeNode`s in the web service model store the `Endpoint`s (in the form of `_AnyEndpoints`). Now, the `_AnyEndpoint` protocol is not somehow generic, but it does define several non-encodable properties (response transformers, guards, etc. all of these are function types and therefore cannot be encoded).
- I don't know much about the context in which @eldcn's code would be running (is it running directly inside Apodini (i.e. the same process & address space), or is it running in a separate process?)
  - If it is running in the same process, tbh I don't see why anything would need to be `Codable` in the first place.
  - If it is running in a different process, and you need the `Codable` conformance in order to "access" the web service's model from a process different to the one in which Apodini is running, what I ended up doing was that I defined [a couple of `Codable` structs](https://github.com/Apodini/Apodini/blob/0f9d7cb794042c8460e88f9e225b3ee0a4a0bd2f/Sources/ApodiniDeployBuildSupport/WebServiceStructure.swift) for collecting the subset of information from the web service model I was interested in: the set of endpoints (for which I cared about each `Endpoint`'s static type, handler id, and deployment options) of the web service as well as deployment options.
- With regards to the "dedicated exporter" point (this is still assuming that @eldcn's code is running out-of-process), I think it would be a really interesting idea to have one exporter which doesn't export any client-accessible endpoints or services or anything, but instead is used by Apodini-extending components (such as eg ApodiniDeploy) to get a "copy" of the structure of the web service. This would also make it a lot easier to define such components. What I'm envisioning here is some kind of "I have the file path and SPM target name of some Apodini web service, and I want to get a data structure containing as much information as possible about the web service as possible" API, kinda like what the [`DeploymentProvider` protocol currently implements](https://github.com/Apodini/Apodini/blob/0f9d7cb794042c8460e88f9e225b3ee0a4a0bd2f/Sources/ApodiniDeployBuildSupport/DeploymentProvider.swift#L94)


@PSchmiedmayer also regarding the number of exporters: i'd guess that in most cases, if you're using the REST exporter, chances are you also want the OpenAPI interface exporter (so there's an implication `REST → OpenAPI`), but I can't really see why someone would want to use the OpenAPI exporter w/out also using the REST exporter (so we also have `OpenAPI !→ REST`.
What I'm getting at here is that I think, if reducing the number of interface exporters is the goal, we could merge the REST and OpenAPI interface exporters into one (eg the existing REST IE), and add an `includeOpenAPIDef` option on that.

Reply:
We will have to discuss the options to make the core data structures `Codable`, but as Lukas also said, it is impossible currently. @lukaskollmer I need to persist all interface related infos of the endpoints and compare different versions, thats why I need `Codable` and the core context runs in the same process that Apodini runs